iosevka: 3.2.2 -> 3.7.1

AluisioASG opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Both packages were outdated and out of sync with one another.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

midchildan commented on pkgs/data/fonts/iosevka/bin.nix:
```diff
@@ -1,7 +1,7 @@
 
 let
-  version = "3.4.6";
+  version = "3.6.2";
```
Do `iosevka` and `iosevka-bin` always have to be in sync with each other? If so, I think we should ensure they'd always stay the same by changing this line to `version = iosevka.version;`.

AluisioASG replied:
They don't need to be, it just happens that `iosevka-bin` gets updated automatically, while `iosevka` needs the `package.json` to be regenerated. It'd make it less confusing if they were updated in lockstep though.

midchildan commented on pkgs/data/fonts/iosevka/package.json:
```diff
@@ -1,28 +1,27 @@
 {
```
Could you add a script that generates this automatically? Here's a working example:

```bash
#!/usr/bin/env bash

# Script to generate package.json needed to create the build environment.

set -eu -o pipefail
cd "$(dirname "${BASH_SOURCE[0]}")"

SRC="$(nix-build -A iosevka.src --no-out-link ../../../..)"
JQ_BIN="$(nix-build -A jq --no-out-link ../../../..)/bin/jq"


main() {
    jq '{
            "name": "iosevka-build-deps",
            "version": .version,
            "scripts": { "build": .scripts.build },
            "dependencies": .dependencies
        }' \
        "$SRC/package.json" \
        > package.json
}

jq() {
    command "$JQ_BIN" "$@"
}

main
```

AluisioASG replied:
The latest push includes an `update.sh` script that does it.

AluisioASG commented:
Made some changes:
- centralized version number and hashes in srcs.nix
- filtered package.json to keep only name, version, and dependencies
- added an update script

midchildan reviewed (changes requested):
Great work! There's a single typo that needs fixing, but otherwise looks good.

AluisioASG commented on pkgs/data/fonts/iosevka/update.sh:
```diff
@@ -0,0 +1,43 @@
+update_srcs "$new_version" "$src_hash" "$bin_hash"
+
+# Rebuild package.json for the new version.
+package_json=$(nix-build "$nixpkgs" --no-out-link -a iosevka.src)/package.json
```
Thanks!

rileyinman reviewed (approved):
Sorry it took me a bit to get around to this—the changes look great. The auto-update script especially will be handy, since the update process was a bit tedious and that's one of the reasons I hadn't gotten around to the new version yet. Thanks for the improvements!

AluisioASG commented:
Updated to 3.7.0.

@ryantm what's the current procedure for updating node-packages.nix? I've removed the update from this PR but I'm guessing it'll break the ofborg CI run (as well as the automated update once I stop calling generate.sh from the update script).

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/call-for-nodepackages-maintainers/9176/28

doronbehar commented on pkgs/data/fonts/iosevka/update.sh:
```diff
@@ -0,0 +1,43 @@
+
+# Rebuild package.json for the new version.
+package_json=$(nix-build "$nixpkgs" --no-out-link -A iosevka.src)/package.json
+jq '{name: "iosevka-build-deps", version, dependencies}' "$package_json" >package.json
```
I think what I get from here is that it's not available on npm? That's why you can't put it in `node-packages.json`? If so, you should know it's possible to do this:

https://github.com/NixOS/nixpkgs/blob/a38590d91ce79e56d1bae79c5824008e906813ef/pkgs/development/node-packages/node-packages.json#L145

And you then will need only to update the version tag there every once in a while.. If that will be possible, you should be able to remove a lot of moving parts from these derivations.

AluisioASG replied:
Well, that does indeed simplify things a lot. I had to use a tarball though, as node2nix doesn't perform a shallow fetch and Iosevka's repo is 1GB+ in size.

AluisioASG commented:
The latest commits are a significant rework based on @doronbehar's suggestion, at the expense of keeping the packages unsynchronized as they are today.

iosevka-bin can be updated with `genericUpdater`, but it needs a `src` attribute pointing to itself for that to work, because of

https://github.com/NixOS/nixpkgs/blob/242441fea2a8981dccf5800288c7ca02a1c31cac/pkgs/common-updater/scripts/update-source-version#L82-L83

iosevka's update script now only needs to change a URL in node-packages.json, but that update isn't propagated to the `iosevka` package until node-packages.nix is regenerated.

Since the two packages are again separated, this may be better off as two separate PRs. I'm marking this up as a draft so it doesn't get merged until at least I've cleaned up the commits.

doronbehar commented on pkgs/data/fonts/iosevka/bin.nix:
```diff
@@ -1,28 +1,39 @@
-  version = "3.4.6";
-in fetchzip {
-  name = "iosevka-bin-${version}";
+  pname = "iosevka-bin";
```
What is the purpose of this package, and what does it provide that `iosevka` doesn't? Overall, I'm asking because I'd like to push this towards a minimal PR that only puts `iosevka` into `node-packages.json` to allow @cideM grab it to his next round of `./generate.sh`.

AluisioASG replied:
According to [the PR that inits it](https://github.com/NixOS/nixpkgs/pull/32106) and [one of the comments it points to](https://github.com/NixOS/nixpkgs/pull/31835#issuecomment-347048242), one of Iosevka's dependencies (otfcc) didn't build on aarch64 back then. However, I no longer see otfcc in the dependencies list (possibly replaced with [ot-builder](https://github.com/ot-builder/monorepo)), so it'd be worth checking if it works now, [which it seems like it does?](https://github.com/NixOS/nixpkgs/pull/99004/checks?check_run_id=1333017601)

AluisioASG replied:
Oh, there is also [this](https://github.com/NixOS/nixpkgs/pull/101132) now.

AluisioASG committed:
iosevka-bin: move update script

Move bin.nix's update script to update-bin.sh so it doesn't conflict
with default.nix's update script.

AluisioASG committed:
iosevka: 3.2.2 -> 3.7.1; fetch source through nodePackages

Previously we used nodePackages only to fetch Iosevka's build
dependencies and then fetched the source code ourselves.  Updates
involved changing the version and hashes in the `iosevka` derivation and
then running node-packages/generate.sh to update the build dependencies,
which in turns updates *all* of node-packages.nix.

A new proposed policy for handling node-packages.nix updates involves
batching those updates.  Previously, that would mean `iosevka` and its
build dependencies could end up out of sync until the batched update was
run.

To work with the new policy, we now fetch Iosevka's source code (and
not only its dependencies) through nodePackages.  Updates are done by
changing the source URL in node-packages.json, which eventually becomes
part of an update of node-packages.nix, which is then propagated to
`iosevka` itself.

One con of this strategy is that errors can not be caught directly
after the update, but only after node-packages.nix is regenerated.

AluisioASG commented:
@doronbehar since #101132 took care of `iosevka-bin`, I removed the changes that dealt with it (but I moved its update script to make clear what package it is updating).

Should I drop the last commit (that updates node-packages.nix) even if it breaks `iosevka` on master in the meantime?

doronbehar commented:
OK. It took me a while to grab my head around the changes. It looks good, although it's a bit of a mess. The problem is that many users are probably happy with iosevka-bin since it's easier to compile an override of it if one wishes only to get different / certain variants. Hence it's not suitable to discard the prebuild derivation for the sake of the source build derivation.

Unrelated, for iosevka, it's good to put it's deps in node-packages.json (as it used to be, only previously it was a manually downloaded package.json) so it will (hopefully) get updated frequently by the node packages maintainers (which we currently lack officially).

I was considering maybe suggesting to put the update script of the source build derivation in node-packages' `./generate.sh` but that seems like too much to me, and inappropriate though I'm open to argue about it.

I have a few nitpicks left though (appropriate to put in separate commits in this PR):

```diff
diff --git i/pkgs/data/fonts/iosevka/bin.nix w/pkgs/data/fonts/iosevka/bin.nix
index c3b7767bccc..853c338e184 100644
--- i/pkgs/data/fonts/iosevka/bin.nix
+++ w/pkgs/data/fonts/iosevka/bin.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchurl, unzip
+{ stdenv, lib, fetchurl, unzip, iosevka # Only for it's meta
 , variant ? ""
 }:
 
@@ -29,15 +29,7 @@ in stdenv.mkDerivation rec {
     unzip -d $out/share/fonts/truetype $src
   '';
 
-  meta = with lib; {
-    homepage = "https://be5invis.github.io/Iosevka/";
-    downloadPage = "https://github.com/be5invis/Iosevka/releases";
-    description = ''
-      Slender monospace sans-serif and slab-serif typeface inspired by Pragmata
-      Pro, M+ and PF DIN Mono, designed to be the ideal font for programming.
-    '';
-    license = licenses.ofl;
-    platforms = platforms.all;
+  meta = iosevka.meta // {
     maintainers = [ maintainers.cstrahan ];
   };
 
diff --git i/pkgs/data/fonts/iosevka/default.nix w/pkgs/data/fonts/iosevka/default.nix
index 9ef751688db..ef2df780e83 100644
--- i/pkgs/data/fonts/iosevka/default.nix
+++ w/pkgs/data/fonts/iosevka/default.nix
@@ -75,7 +75,7 @@ stdenv.mkDerivation rec {
 
   buildPhase = ''
     runHook preBuild
-    npm run build --no-update-notifier -- --jCmd=$NIX_BUILD_CORES ttf::$pname >/dev/null
+    npm run build --no-update-notifier -- --jCmd=$NIX_BUILD_CORES ttf::$pname
     runHook postBuild
   '';
 
```

1. Don't silence the output of `npm run build` - it's interesting IMO.
2. Don't repeat yourself (meta information).

doronbehar commented:
> Should I drop the last commit (that updates node-packages.nix) even if it breaks `iosevka` on master in the meantime?

No, we can't break it directly on `master`. If there was an upcoming nodePackages `./generate.sh` update, it would have been appropriate to tell the updater to grab the commits from here and we'd close this PR.

AluisioASG commented:
> Don't silence the output of `npm run build` - it's interesting IMO.

As far as I can tell there is no way to get a useful log without getting progress spam in the way, and any redirects just shut down all logging:

```
> iosevka@3.7.1 build /build/Iosevka-3.7.1
> verda -f verdafile.js "--jCmd=2" "--verbosity=6" "ttf::iosevka"
⢀⠀ ● Note Invalidating build journal due to self-tracking.
⡀⠀  | CPU:000% - RAM:000%⡀⠀ Complete 04 - Executing 2 - Pending 59 (06%) | CPU:000% - RAM:000%⠄⠀ Complete 004 - Executing 0 - Pending 110 (03%) | CPU:000% - RAM:000%⢂⠀ Complete 059 - Executing 2 - Pending 109 (34%) | CPU:000% - RAM:000%⡂⠀ Complete 062 - Executing 2 - Pending 161 (27%) | CPU:000% - RAM:000%⠅⠀ Complete 063 - Executing 0 - Pending 167 (27%) | CPU:000% - RAM:000%♦ node ./node_modules/patel/bin/patel-c --strict font-src/kits/boole-kit.ptl -o
⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%♦ node ./node_modules/patel/bin/patel-c --strict font-src/kits/spiro-kit.ptl -o
⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⢃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⡃⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⠍⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:000% - RAM:000%⢋⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:053%⡋⠀ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%⠍⠁ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%⢋⠁ Complete 068 - Executing 2 - Pending 199 (25%) | CPU:051% - RAM:054%♦ node ./node_modules/patel/bin/patel-c --strict font-src/meta/aesthetics.ptl
^C
```

Even if I reduce verbosity, it'll keep saving the progress bar updates for printing later.

```
> iosevka@3.7.1 build /build/Iosevka-3.7.1
> verda -f verdafile.js "--jCmd=2" "--verbosity=5" "ttf::iosevka"

⢀⠀ ● Note Invalidating build journal due to self-tracking.
```
When that build finished, it produced 7200 lines of the following:
![Progress bar spam after build](https://user-images.githubusercontent.com/1904165/97779350-2ebcd300-1b5c-11eb-9789-accd614f50ae.png)

AluisioASG committed:
iosevka: remove dependency on otfcc

As of Iosevka 3.7.0, otfcc is no longer used.  I haven't checked if
the situation has changed since [2017] but this should make Iosevka
available on aarch64-linux and *-darwin.

[2017]: https://github.com/NixOS/nixpkgs/pull/31835#issuecomment-347048242

AluisioASG committed:
iosevka: add AluisioASG as maintainer

AluisioASG committed:
iosevka: add update script

As outlined in the previous commit, we just need to update the source
URL in node-packages.json, and wait for node-packages.nix to be rebuilt
from it.

AluisioASG commented:
Sharing `meta` seems like it'll mess up with iosevka-bin's update script, as it locates the file to update with

https://github.com/NixOS/nixpkgs/blob/f28c4a120f5cacdad318782ccdd5b3599cea7e65/pkgs/data/fonts/iosevka/update.sh#L16

and what I get when doing `meta = iosevka.meta // ...` is

```
nix-repl> iosevka-bin.meta.position
"/home/aasg/nix/nixpkgs/iosevka/pkgs/data/fonts/iosevka/default.nix:98"
```

@deviant is there a problem if I replace that line with the following, or similar (I see you also write to variants.nix further down)?

```sh
file=$(dirname "${BASH_SOURCE[0]}")/bin.nix
```

AluisioASG commented:
Meanwhile, I removed otfcc from the dependencies, added myself as maintainer, and updated the nodePackages update.

deviant commented:
> Sharing `meta` seems like it'll mess up with iosevka-bin's update script, as it locates the file to update with
> 
> https://github.com/NixOS/nixpkgs/blob/f28c4a120f5cacdad318782ccdd5b3599cea7e65/pkgs/data/fonts/iosevka/update.sh#L16
> 
> @deviant is there a problem if I replace that line with the following, or similar (I see you also write to variants.nix further down)?
> 
> ```shell
> file=$(dirname "${BASH_SOURCE[0]}")/bin.nix
> ```

Not by me. I pretty much just copied that out of the common updater script. Perhaps it would be good to move iosevka-bin into a directory of its own, however? That way the file names can be more normal.

AluisioASG commented:
> Perhaps it would be good to move iosevka-bin into a directory of its own, however? That way the file names can be more normal.

Possibly. The packages have little in common by now and we already do that with e.g. Firefox.

doronbehar commented:
Sharing the meta should not be blocked by the design of the update script. You can put the version in a separate file and not use `nix-instantiate` or whatever it is.

@AluisioASG I'm sorry this got held up, but now we have a merge conflict...

If you could please change the update script to not rely on `meta.position` and use a safer way to increment the version, it'd be great.

doronbehar commented:
As for the merge conflict, I don't know what to do. It's indeed frustrating. We need to coordinate runs of node's `./generate.sh`... (cc @cideM).

doronbehar commented:
We could also (after the version increment issue will be solved) kindly ask @cideM to grab the commits from here (and perhaps along [this](https://github.com/NixOS/nixpkgs/commit/db7a346d485b890fa68424880fb88368d1615cdf) commit [please](https://discourse.nixos.org/t/call-for-nodepackages-maintainers/9176/27) :pray:) and then put them both in a separate PR and then run `./generate.sh`.

AluisioASG committed:
iosevka-bin: Use iosevka's meta

The two packages have the same metadata, except for the maintainers.

AluisioASG commented:
I changed the update script, updating nodePackages now, should be done within the hour.

AluisioASG committed:
nodePackages: update

doronbehar reviewed (approved):
Well done.

deviant commented on pkgs/data/fonts/iosevka/update-bin.sh:
```diff
@@ -1,20 +1,20 @@
 release=$(curl -s https://api.github.com/repos/be5invis/Iosevka/releases/latest)
 
-oldVersion=$(nix-instantiate --eval -E 'with import ./. {}; lib.getVersion iosevka-bin' | tr -d '"')
+oldVersion=$(nix-instantiate --eval -E 'with import ../../../.. {}; lib.getVersion iosevka-bin' | tr -d '"')
```
This seems wrong; you're supposed to run the update script from the root directory of nixpkgs.

doronbehar replied:
The script changes the directory of itself before executing these commands.

deviant replied:
It seems, at least, uncommon:

```
v@february ~/s/nixpkgs (local)> rg -t sh 'import \./' | wc -l
10
v@february ~/s/nixpkgs (local)> rg -t sh 'import \.\./' | wc -l
3
```

deviant commented on pkgs/data/fonts/iosevka/update-default.sh:
```diff
@@ -0,0 +1,21 @@
+set -euo pipefail
+cd "$(dirname "${BASH_SOURCE[0]}")"
+
+nixpkgs=../../../..
```
Same here.

deviant commented:
You'll note that the [nerdfonts update script you mentioned](https://github.com/NixOS/nixpkgs/blob/a7f375d3354af159eab45896e012fb9e856b93cb/pkgs/data/fonts/nerdfonts/update.sh) does not actually cd into its directory.
